pkg/ring/ring.go

```diff
@@ -126,13 +127,15 @@ func New(cfg Config) (*Ring, error) {
 
 // Stop the distributor.
 func (r *Ring) Stop() {
-	close(r.quit)
+	r.quit()
 	<-r.done
 }
 
 func (r *Ring) loop() {
+	var ctx context.Context
+	ctx, r.quit = context.WithCancel(context.Background())
```

Comment:
Is there a potential data race here?  Assigning to r.quit from a background goroutine that is.  Doubt it will actually manifest, but I would probably create the context/WithCancel in the constructor and pass in to `loop`.